File: pomodoro.pl
#!/usr/bin/env perl
# A simple command-line Pomodoro timer

use strict;
use warnings;
use POSIX qw(floor);

$| = 1;  # Disable output buffering

my $WORK_DURATION = 25;
my $SHORT_BREAK = 5;
my $LONG_BREAK = 15;
my $POMODOROS_BEFORE_LONG_BREAK = 4;

my $pomodoro_count = 0;

sub format_time {
    my ($seconds) = @_;
    my $mins = floor($seconds / 60);
    my $secs = $seconds % 60;
    return sprintf("%02d:%02d", $mins, $secs);
}

sub countdown {
    my ($duration_minutes, $label) = @_;
    my $total_seconds = $duration_minutes * 60;

    print "\n$label - $duration_minutes minutes\n";
    print "-" x 30 . "\n";

    for (my $remaining = $total_seconds; $remaining > 0; $remaining--) {
        print "\r  Time remaining: " . format_time($remaining) . " ";
        sleep(1);
    }

    print "\r  Time remaining: 00:00 \n";
    print "\n*** $label complete! ***\a\n";
    return 1;
}

sub cleanup {
    print "\n\nPomodoro session ended.\n";
    print "Completed $pomodoro_count pomodoro(s). Great work!\n";
    exit(0);
}

$SIG{INT} = \&cleanup;

print "=" x 40 . "\n";
print "       POMODORO TIMER\n";
print "=" x 40 . "\n";
print "  Work session:  $WORK_DURATION min\n";
print "  Short break:   $SHORT_BREAK min\n";
print "  Long break:    $LONG_BREAK min (every $POMODOROS_BEFORE_LONG_BREAK pomodoros)\n";
print "=" x 40 . "\n";
print "Press Ctrl+C to stop\n\n";

while (1) {
    $pomodoro_count++;
    print "\n[Pomodoro #$pomodoro_count]\n";

    countdown($WORK_DURATION, "WORK");

    print "Press Enter to start break...";
    <STDIN>;

    if ($pomodoro_count % $POMODOROS_BEFORE_LONG_BREAK == 0) {
        countdown($LONG_BREAK, "LONG BREAK");
    } else {
        countdown($SHORT_BREAK, "SHORT BREAK");
    }

    print "Press Enter to start next pomodoro...";
    <STDIN>;
}

File: pomodoro.lua
#!/usr/bin/env lua
-- A simple command-line Pomodoro timer in Lua

local function format_time(seconds)
    local mins = math.floor(seconds / 60)
    local secs = seconds % 60
    return string.format("%02d:%02d", mins, secs)
end

local function countdown(duration_minutes, label)
    local total_seconds = duration_minutes * 60
    print(string.format("\n%s - %d minutes", label, duration_minutes))
    print(string.rep("-", 30))

    for remaining = total_seconds, 1, -1 do
        io.write(string.format("\r  Time remaining: %s ", format_time(remaining)))
        io.flush()

        -- Sleep for 1 second (using os.execute with platform-specific sleep)
        local sleep_cmd
        if package.config:sub(1,1) == '\\' then
            -- Windows
            sleep_cmd = "timeout /t 1 /nobreak >nul 2>&1"
        else
            -- Unix-like (Linux, macOS)
            sleep_cmd = "sleep 1"
        end
        os.execute(sleep_cmd)
    end

    io.write("\r  Time remaining: 00:00 \n")
    print(string.format("\n*** %s complete! ***\a", label))
    return true
end

local function pomodoro()
    local work_duration = 25
    local short_break = 5
    local long_break = 15
    local pomodoros_before_long_break = 4

    print(string.rep("=", 40))
    print("     🍅 POMODORO TIMER (Lua)")
    print(string.rep("=", 40))
    print(string.format("  Work session:  %d min", work_duration))
    print(string.format("  Short break:   %d min", short_break))
    print(string.format("  Long break:    %d min (every %d pomodoros)",
          long_break, pomodoros_before_long_break))
    print(string.rep("=", 40))
    print("Press Ctrl+C to stop\n")

    local pomodoro_count = 0

    while true do
        pomodoro_count = pomodoro_count + 1
        print(string.format("\n[Pomodoro #%d]", pomodoro_count))

        -- Work session
        if not countdown(work_duration, "WORK") then
            break
        end

        io.write("Press Enter to start break...")
        io.read()

        -- Break session
        local is_long_break = (pomodoro_count % pomodoros_before_long_break == 0)
        if is_long_break then
            if not countdown(long_break, "LONG BREAK") then
                break
            end
        else
            if not countdown(short_break, "SHORT BREAK") then
                break
            end
        end

        io.write("Press Enter to start next pomodoro...")
        io.read()
    end

    print(string.format("\nCompleted %d pomodoro(s). Great work! 🎉", pomodoro_count))
end

-- Run the pomodoro timer
pomodoro()
